Karl/claire.h

```diff
@@ -0,0 +1,174 @@
+
+  static std::string GenerateRandomCodename() {
+    std::string codename;
+    for (int i = 0; i < 6; ++i) {
+      codename += static_cast<char>(current::random::CSRandomInt('A', 'Z'));
+    }
+    return codename;
+  }
+
+  GenericClaire(Locator karl, const std::string& service, uint16_t port)
+      : register_called_(false),
+        registered_(false),
+        karl_(karl),
+        service_(service),
+        codename_(GenerateRandomCodename()),
+        port_(port),
+        us_start_(current::time::Now()),
+        http_scope_(HTTP(port).Register("/.current",
+                                        [this](Request r) {
+                                          const auto& qs = r.url.query;
+                                          const bool all = qs.has("all") || qs.has("a");
+                                          const bool build = qs.has("build") || qs.has("b");
+                                          if (!all && build) {
+                                            r(build::Info());
+                                          } else {
+                                            if (!registered_) {
+                                              ClaireStatusBase response;
+                                              FillBase(response, all);
+                                              r(response);
+                                            } else {
+                                              status_t response;
+                                              FillBase(response, all);
+                                              if (status_generator_) {
+                                                std::lock_guard<std::mutex> lock(mutex_);
+                                                response.runtime = status_generator_();
+                                              }
+                                              r(response);
+                                            }
+                                          }
+                                        })) {
+    // TODO(dkorolev) + TODO(mzhurovich): Should Claire at least check whether Karl is available at
+    // construction?
+  }
+
+  ~GenericClaire() {
+    if (keepalive_thread_.joinable()) {
+      keepalive_thread_.join();
+    }
+  }
+
+  void Register(status_generator_t status_filler = nullptr, bool strict = false) {
+    // Register this Claire with Karl and spawn the thread to send regular keepalives.
+    // If `strict` is true, throw if Karl can be not be reached.
+    // If `strict` is false, just start the keepalives thread.
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (!register_called_) {
+      register_called_ = true;
+      status_generator_ = status_filler;
+      // Only register once.
+      if (strict) {
+        // During this call, Karl would crawl the endpoint of this service, and, if everything is successful,
+        // register this service as the running and browsable one.
```

Comment:
I still struggle to see what's wrong with registering `.current` in Claire's constructor.

If the user does nothing, their binary would still report its service/codename/port/time skew/uptime/build_info. All of the above are both straightforward to generate/respond with, and useful for monitoring/development/babysitting purposes.

As we move to scoped `HTTP(port)` ownership, I'd expect each class claiming (exclusive) access to certain HTTP port would claim it first thing in the constructor ("fail early, fail better"). Upon making that the new standard, I see no reason to not move Claire initialization right into there, effectively by designing it as `class HTTPPortAccessor { Register(...)  ... };` and `class Claire : HTTPPortAccessor { ... }`, so the user could use `Claire` to first register HTTP endpoints, and then `.RegisterWithKarl()` itself.

Let's converge on something in the next 36 hours :-)